keps/sig-autoscaling/2021-scale-from-zero/README.md

```diff
@@ -0,0 +1,681 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+From an UX perspective the two stage opt-out / opt-in from scale to zero might feel a bit tedious, but the only other available option seems to be deprecating the implicit HPA pause on `replicas: 0`. While this might provide an improved
+UX, it would require a full deprecation cycle (12 months) before graduating this feature from alpha to beta.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We would add `EnableScaleToZero *bool` to the HPA `spec.behavior`.
+
+```golang
+type HorizontalPodAutoscalerBehavior struct {
+    ScaleUp           *HPAScalingRules
+    ScaleDown         *HPAScalingRules
+    EnableScaleToZero *bool
+}
+
+type HorizontalPodAutoscalerSpec struct {
+    ScaleTargetRef CrossVersionObjectReference
+    MinReplicas    *int32
+    MaxReplicas    int32
+    Metrics        []MetricSpec
+    Behavior       *HorizontalPodAutoscalerBehavior
+}
+```
+
+The `EnableScaleToZero` controls whether the `MinReplicas` can set to `>=0` (`true`, new behavior) or `>=1` (`false`, current behavior). The default will be `false` to preserve the current behavior.
+
+If `EnableScaleToZero` has been enabled, it can only be disabled when the scaled resource has at least one replica
+running and `MinReplicas` is `>=1`.
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+Most logic related to this KEP is contained in the HPA controller so the testing of
+the various `minReplicas`, `replicas` and `enableScaleToZero` should be achievable with unit tests.
+
+Additionally integration tests should be added for enable scale to zero by, setting
+ `enableScaleToZero: true`, setting `enableScaleToZero: true` and waiting for `replicas` to become `0` and another test for increasing `minReplicas: 1`, setting `enableScaleToZero: false` and observing that `replicas` became `1` again.
+
```

Comment:
something that occurs to me while re-reading this section, what happens if the user changes `enableScaleToZero` while either `replicas` or `minreplicas` has been set to `0`. would there be any need to test these type of failure states?

Reply:
I think those cases should be covered by controller unit tests. Do you think it is worth explicitly calling them out here?

Reply:
We should call them out here.  This test plan is for the whole feature, so we should list tests which should be implemented at every level (api, admission, controller, e2e).

Reply:
After merging this KEP we should come back the fill out the test plan a bit. Just to make sure we haven't missed anything.